### ISSUE TITLE
Timeout performing SETEX 

### ISSUE THREAD

**Issue by Xhanti** [StackExchange/StackExchange.Redis#270]
Hi,

We periodically keep getting exceptions of this nature:

```
Timeout performing SETEX HybridCache_Tuple`2_21.110.24.0_OutputCache\Shared\RootHeader_Tuple`2_21... EXCEPTION: System.TimeoutException
Timeout performing SETEX HybridCache_Tuple`2_21.110.24.0_OutputCache\Shared\RootHeader_Tuple`2_21.110.24.0_h3jO0rZr0CdtQG2oJxayyNhykQl15sJU2fOwRHbVGTM=, inst: 2, mgr: ExecuteSelect, queue: 3, qu: 0, qs: 3, qc: 0, wr: 0, wq: 0, in: 5, ar: 0, IOCP: (Busy=0,Free=1000,Min=8,Max=1000), WORKER: (Busy=16,Free=32751,Min=8,Max=32767)

at StackExchange.Redis.ConnectionMultiplexer.ExecuteSyncImpl[T](Message message, ResultProcessor`1 processor, ServerEndPoint server) in c:\TeamCity\buildAgent\work\3ae0647004edff78\StackExchange.Redis\StackExchange\Redis\ConnectionMultiplexer.cs:line 1922
at StackExchange.Redis.RedisBase.ExecuteSync[T](Message message, ResultProcessor`1 processor, ServerEndPoint server) in c:\TeamCity\buildAgent\work\3ae0647004edff78\StackExchange.Redis\StackExchange\Redis\RedisBase.cs:line 80
at StackExchange.Redis.RedisDatabase.StringSet(RedisKey key, RedisValue value, Nullable`1 expiry, When when, CommandFlags flags) in c:\TeamCity\buildAgent\work\3ae0647004edff78\StackExchange.Redis\StackExchange\Redis\RedisDatabase.cs:line 1545
at Korbitec.Property24.RedisCache`1.c__DisplayClass4.b__3() in c:\Work\Portal\portal-stable\Property24\Cache\CacheTypes\RedisCache.cs:line 42
at Korbitec.Property24.RedisCache`1.TimeoutExceptionHandledAction(Action action) in c:\Work\Portal\portal-stable\Property24\Cache\CacheTypes\RedisCache.cs:line 85

Source: StackExchange.Redis
Target Site: T ExecuteSyncImpl[T](StackExchange.Redis.Message, StackExchange.Redis.ResultProcessor`1[T], StackExchange.Redis.ServerEndPoint)
```

##### Exception Data
```
redis-command , SETEX HybridCache_Tuple`2_21.110.24.0_OutputCache\Shared\RootHeader_Tuple`2_21.110.24.0_h3jO0rZr0CdtQG2oJxayyNhykQl15sJU2fOwRHbVGTM=
redis-server , drptlredis1:6379
Redis-Message , SETEX HybridCache_Tuple`2_21.110.24.0_OutputCache\Shared\RootHeader_Tuple`2_21.110.24.0_h3jO0rZr0CdtQG2oJxayyNhykQl15sJU2fOwRHbVGTM=
Redis-Instantaneous , 2
Redis-Manager-State , ExecuteSelect
Redis-Queue-Length , 3
Redis-Queue-Outstanding , 0
Redis-Queue-Awaiting-Response , 3
Redis-Queue-Completion-Outstanding , 0
Redis-Active-Writers , 0
Redis-Write-Queue , 0
Redis-Inbound-Bytes , 5
Redis-Active-Readers , 0
**Redis-ThreadPool-IO-Completion , (Busy=0,Free=1000,Min=8,Max=1000)**
Redis-ThreadPool-Workers , (Busy=16,Free=32751,Min=8,Max=32767) 
```

Is the line in bold under Exception data expected?

Regards,
Xhanti

**Comment by NickCraver** [comment]
@Xhanti the bolded line means that no IO completion port threads are active processing responses, which makes sense because you timed out waiting for responses to arrive.

What's your setup here? What kind of infrastructure/latency are we looking at, and what are your timeout set to? Also check the server's `SLOWLOG` to see if redis itself stalled due to some long-running operation when your timeouts occurred.

**Comment by Xhanti** [comment]
HI @NickCraver ,

Firstly thank you very much for the response.

We have 5 nodes speaking to one master and slave, with sentinel on the Redis instances and running on a third box. The nodes are configured to only read from and write to the master.
The Redis boxes run on Linux, version 2.8.4. The sync time out is 5 seconds and response time out is not specified (should they differ and if so by how much?). The slow log has nothing when we check. 

This is all on our internal network which is quite fast, the specifics escape me(I'll need to email operations for specifics). 

If there is anything else you need, please don't hesitate to comment .

Regards,
Xhanti

**Comment by Xhanti** [comment]
Hi @NickCraver ,

Spoke to Operations, we have a minimum 10'Gigabit connection between vm's.

Checked the slowlog and we got this:

127.0.0.1:6379> slowlog get 5
1) 1) (integer) 28678
   2) (integer) 1453877682
   3) (integer) 12713
   4) 1) "EVALSHA"
      2) "6009a2dbd94762afd58de589a21a47b27c82d14b"
      3) "3"
      4) "_rv52o4ucfcdm1haswtoon22b_Write_Lock"
      5) "_rv52o4ucfcdm1haswtoon22b_Data"
      6) "_rv52o4ucfcdm1haswtoon22b_Internal"
2) 1) (integer) 28677
   2) (integer) 1453877682
   3) (integer) 11794
   4) 1) "EVALSHA"
      2) "6009a2dbd94762afd58de589a21a47b27c82d14b"
      3) "3"
      4) "_rv52o4ucfcdm1haswtoon22b_Write_Lock"
      5) "_rv52o4ucfcdm1haswtoon22b_Data"
      6) "_rv52o4ucfcdm1haswtoon22b_Internal"
3) 1) (integer) 28676
   2) (integer) 1453877682
   3) (integer) 13893
   4) 1) "EVALSHA"
      2) "6009a2dbd94762afd58de589a21a47b27c82d14b"
      3) "3"
      4) "_rv52o4ucfcdm1haswtoon22b_Write_Lock"
      5) "_rv52o4ucfcdm1haswtoon22b_Data"
      6) "_rv52o4ucfcdm1haswtoon22b_Internal"
4) 1) (integer) 28675
   2) (integer) 1453877682
   3) (integer) 12167
   4) 1) "EVALSHA"
      2) "6009a2dbd94762afd58de589a21a47b27c82d14b"
      3) "3"
      4) "_rv52o4ucfcdm1haswtoon22b_Write_Lock"
      5) "_rv52o4ucfcdm1haswtoon22b_Data"
      6) "_rv52o4ucfcdm1haswtoon22b_Internal"
5) 1) (integer) 28674
   2) (integer) 1453877681
   3) (integer) 11530
   4) 1) "EVALSHA"
      2) "6009a2dbd94762afd58de589a21a47b27c82d14b"
      3) "3"
      4) "_rv52o4ucfcdm1haswtoon22b_Write_Lock"
      5) "_rv52o4ucfcdm1haswtoon22b_Data"
      6) "_rv52o4ucfcdm1haswtoon22b_Internal"

**Comment by NickCraver** [comment]
@Xhanti okay so that's a LUA script running and stalling redis (which runs these single threaded). The time entries here are unix epoch times ([handy converter](http://www.epochconverter.com/)):
> 2) (integer) 1453877682

Do these stalls line up with your timeouts exceptions? If so the problem lies in that script tying up redis for too long.

**Comment by Xhanti** [comment]
Thanks @NickCraver , 
I"ll check it and get back to you.

Regards

**Comment by Xhanti** [comment]
Hi @NickCraver, 

After much toil and trouble we realized that our redis server was constantly evicting(and not expiring) entries indicating that our redis server was under constant abnormal stress. We compressed our values in the cache and dropped our evictions to zero. Since then we have not had a timeout. Thanks again for the quick response.

I'm adding this now in the hopes it can help someone else.
Closing this issue now.

Regards
